src/librbd/AioRequest.cc

```diff
@@ -52,37 +57,160 @@ namespace librbd {
 	     m_parent_completion, 0);
   }
 
+  static inline bool is_cor(ImageCtx *ictx, librados::snap_t snap_id) {
+     return (ictx->cct->_conf->rbd_clone_copy_on_read) &&
+             (!ictx->read_only) && (snap_id == CEPH_NOSNAP);
+  }
+
   /** read **/
 
+  void AioRead::guard_read()
+  {
+    RWLock::RLocker l(m_ictx->snap_lock);
+    RWLock::RLocker l2(m_ictx->parent_lock);
+
+    vector<pair<uint64_t,uint64_t> > image_extents;
+    Striper::extent_to_file(m_ictx->cct, &m_ictx->layout,
+                            m_object_no, 0, m_ictx->layout.fl_object_size,
+                            image_extents);
+
+    uint64_t image_overlap = 0;
+    m_ictx->get_parent_overlap(m_snap_id, &image_overlap);
+    uint64_t object_overlap =
+      m_ictx->prune_parent_extents(image_extents, image_overlap);
+    if (object_overlap) {
+      ldout(m_ictx->cct, 20) << __func__ << " guarding read" << dendl;
+      m_state = LIBRBD_AIO_READ_GUARD;
+    }
+  }
+
   bool AioRead::should_complete(int r)
   {
-    ldout(m_ictx->cct, 20) << "should_complete " << this << " " << m_oid << " " << m_object_off << "~" << m_object_len
-			   << " r = " << r << dendl;
+    ldout(m_ictx->cct, 20) << "AioRead::should_complete " << this
+                           << " " << m_oid << " " << m_object_off
+                           << "~" << m_object_len << " r = " << r
+                           << " cor = " << is_cor(m_ictx, m_snap_id)
+                           << " readonly = " << m_ictx->read_only
+                           << dendl;
 
-    if (!m_tried_parent && r == -ENOENT) {
-      RWLock::RLocker l(m_ictx->snap_lock);
-      RWLock::RLocker l2(m_ictx->parent_lock);
+    bool finished = true;
 
-      // calculate reverse mapping onto the image
-      vector<pair<uint64_t,uint64_t> > image_extents;
-      Striper::extent_to_file(m_ictx->cct, &m_ictx->layout,
-			    m_object_no, m_object_off, m_object_len,
-			    image_extents);
+    switch (m_state) {
+    case LIBRBD_AIO_READ_GUARD:
+      ldout(m_ictx->cct, 20) << "should_complete " << this
+                             << " READ_CHECK_GUARD" << dendl;
+
+      // This is the step to read from parent
+      if (!m_tried_parent && r == -ENOENT) {
+        RWLock::RLocker l(m_ictx->snap_lock);
+        RWLock::RLocker l2(m_ictx->parent_lock);
+
+        // calculate reverse mapping onto the image
+        vector<pair<uint64_t,uint64_t> > image_extents;
```

Comment:
Minor: you could cache the results from ```AioRead::guard_read```'s invocation of ```Striper::extent_to_file``` and then you would only need to run ```prune_parent_extents``` since the overlap can only get smaller.